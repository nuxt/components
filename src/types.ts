export interface Component {
  pascalName: string
  kebabName: string
  import: string
  asyncImport: string
  export: string
  filePath: string
  shortPath: string
  isAsync?: boolean
  chunkName: string
  /** @deprecated */
  global: boolean
  level: number
  prefetch: boolean | number
  preload: boolean | number
  magicComments: string
}

export interface ScanDir {
  path: string
  pattern?: string | string[]
  ignore?: string[]
  prefix?: string
  isAsync?: boolean
  /** @deprecated */
  global?: boolean | 'dev'
  pathPrefix?: boolean
  level?: number
  prefetch?: boolean | number
  preload?: boolean | number
  extendComponent?: (component: Component) => Promise<Component | void> | (Component | void)
}

export interface ComponentsDir extends ScanDir {
  watch?: boolean
  extensions?: string[]
  transpile?: 'auto' | boolean
}

export type componentsDirHook = (dirs: ComponentsDir[]) => void | Promise<void>
export type componentsExtendHook = (components: (ComponentsDir | ScanDir)[]) => void | Promise<void>

export interface Options {
  dirs: (string | ComponentsDir)[]
  loader: Boolean
}

declare module '@nuxt/types/config/index' {
  interface NuxtOptions {
    components: boolean | Options | Options['dirs']
  }
}

declare module '@nuxt/types/config/hooks' {
  interface NuxtOptionsHooks {
    'components:dirs'?: componentsDirHook
    'components:extend'?: componentsExtendHook
    components?: {
      dirs?: componentsDirHook
      extend?: componentsExtendHook
    }
  }
}
