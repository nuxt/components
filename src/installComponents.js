global.installComponents = function (c, l) {
  const t = typeof c.exports === 'function'
  const o = t ? c.exports.extendOptions : c.options

  o.components = (t ? c.exports.options.components : o.components) || {}

  for (var i in l) {
    o.components[i] = o.components[i] || l[i]
  }
}
