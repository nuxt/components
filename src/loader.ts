import { createUnplugin } from 'unplugin'
import MagicString from 'magic-string'
import { pascalCase } from 'scule'
import { createFilter } from '@rollup/pluginutils'
import type { FilterPattern } from '@rollup/pluginutils'
import { Component } from './types'

export const DISABLE_COMMENT = '/* nuxt-components disabled */'

export interface Options {
  findComponent(name: string): Component | void | Promise<Component | void>
  include?: FilterPattern
  exclude?: FilterPattern
}

export const loader = createUnplugin<Options>((options) => {
  const filter = createFilter(
    options?.include || [/\.vue$/, /\.vue\?vue/],
    options?.exclude || [/node_modules/, /\.git/, /\.nuxt/]
  )

  return {
    name: 'nuxt-components-loader',
    enforce: 'post',

    transformInclude (id) {
      return filter(id)
    },

    async transform (code, id) {
      if (code.includes(DISABLE_COMMENT)) {
        return
      }

      const s = new MagicString(code)

      let no = 0
      const componentPaths: string[] = []
      const prepend: string[] = []

      for (const match of code.matchAll(/_c\([\s\n\t]*['"](.+?)["']([,)])/g)) {
        const [full, matchedName, append] = match

        if (match.index != null && matchedName && !matchedName.startsWith('_')) {
          const start = match.index
          const end = start + full.length
          const name = pascalCase(matchedName)
          componentPaths.push(name)
          const component = await options!.findComponent(name)
          if (component && !id.startsWith(component.filePath)) {
            const varName = `__nuxt_components_${no}`
            prepend.push(`import ${varName} from "${component.filePath}"`)
            no += 1
            s.overwrite(start, end, `_c(${varName}${append}`)
          }
        }
      }

      if (!prepend.length) {
        return null
      }

      s.prepend(`${DISABLE_COMMENT}${prepend.join(';')};`)
      return {
        code: s.toString(),
        map: s.generateMap({
          source: id,
          includeContent: true
        })
      }
    }
  }
})
