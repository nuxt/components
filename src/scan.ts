import { basename, extname, join, dirname, relative } from 'upath'
import globby from 'globby'
import { camelCase, kebabCase, upperFirst } from 'lodash'
import type { ScanDir, Component } from './types'

const LAZY_PREFIX = 'lazy'
const pascalCase = (str: string = '') => {
  const isFirstCharUppercase = str[0] === str.toUpperCase()[0]
  const containsHyphens = str.includes('-')
  const shouldTransformToPascal = !isFirstCharUppercase || containsHyphens

  return shouldTransformToPascal ? upperFirst(camelCase(str)) : str
}

function sortDirsByPathLength ({ path: pathA }: ScanDir, { path: pathB }: ScanDir): number {
  return pathB.split(/[\\/]/).filter(Boolean).length - pathA.split(/[\\/]/).filter(Boolean).length
}

function prefixComponent (prefix: string = '', { pascalName, kebabName, ...rest }: Component): Component {
  return {
    pascalName: pascalName.startsWith(prefix) ? pascalName : pascalCase(prefix) + pascalName,
    kebabName: kebabName.startsWith(prefix) ? kebabName : `${kebabCase(prefix)}-${kebabName}`,
    ...rest
  }
}

export async function scanComponents (dirs: ScanDir[], srcDir: string): Promise<Component[]> {
  const components: Component[] = []
  const filePaths = new Set<string>()
  const scannedPaths: string[] = []

  for (const { path, pattern, ignore = [], prefix, extendComponent, global, level } of dirs.sort(sortDirsByPathLength)) {
    const resolvedNames = new Map<string, string>()

    for (const _file of await globby(pattern!, { cwd: path, ignore })) {
      const filePath = join(path, _file)

      if (scannedPaths.find(d => filePath.startsWith(d))) {
        continue
      }

      if (filePaths.has(filePath)) { continue }
      filePaths.add(filePath)

      // Resolve componentName
      let componentName = pascalCase(basename(filePath, extname(filePath)).replace(/^\//g, ''))
      const pathPrefix = pascalCase(relative(path, dirname(filePath)).replace(/^\//g, ''))
      const parentDirName = pascalCase(basename(dirname(filePath)))

      if (['Index', parentDirName].includes(componentName)) {
        componentName = pathPrefix
      } else if (!componentName.startsWith(pathPrefix.replace(/s$/, ''))) {
        componentName = pathPrefix + componentName
      }

      if (resolvedNames.has(componentName)) {
        // eslint-disable-next-line no-console
        console.warn(`Two component files resolving to the same name \`${componentName}\`:\n` +
          `\n - ${filePath}` +
          `\n - ${resolvedNames.get(componentName)}`
        )
        continue
      }
      resolvedNames.set(componentName, filePath)

      const pascalName = pascalCase(componentName)
      const kebabName = kebabCase(componentName)
      const shortPath = filePath.replace(srcDir, '')
      const chunkName = 'components/' + kebabName

      let _c = prefixComponent(prefix, {
        filePath,
        pascalName,
        kebabName,
        chunkName,
        shortPath,
        import: '',
        asyncImport: '',
        export: 'default',
        global: Boolean(global),
        level: Number(level)
      })

      if (typeof extendComponent === 'function') {
        _c = (await extendComponent(_c)) || _c
      }

      const _import = _c.import || `require('${_c.filePath}').${_c.export}`
      const _asyncImport = _c.asyncImport || `function () { return import('${_c.filePath}' /* webpackChunkName: "${_c.chunkName}" */).then(function(m) { return m['${_c.export}'] || m }) }`

      const component = {
        ..._c,
        import: _import
      }
      const lazyComponent = prefixComponent(LAZY_PREFIX, {
        ..._c,
        async: true,
        import: _asyncImport
      })

      // Check if component is already defined, used to overwite if level is inferiour
      const definedComponent = components.find(c => c.pascalName === component.pascalName)
      if (definedComponent && component.level < definedComponent.level) {
        Object.assign(definedComponent, component)
        const definedLazyComponent = components.find(c => c.pascalName === lazyComponent.pascalName)
        definedLazyComponent && Object.assign(definedLazyComponent, lazyComponent)
      } else if (!definedComponent) {
        components.push(component)
        components.push(lazyComponent)
      }
    }

    scannedPaths.push(path)
  }

  return components
}

export function matcher (tags: string[], components: Component[]) {
  return tags.reduce((matches, tag) => {
    const match = components.find(({ pascalName, kebabName }) => [pascalName, kebabName].includes(tag))
    match && matches.push(match)
    return matches
  }, [] as Component[])
}
