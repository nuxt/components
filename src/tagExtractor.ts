
import { readFileSync } from 'fs'
import {
  compile,
  parseComponent,
  ModuleOptions as CompilerModuleOptions
} from 'vue-template-compiler'

export async function extractTags (resourcePath: string): Promise<string[]> {
  const tags = new Set<string>()
  const file = (await readFileSync(resourcePath)).toString('utf8')
  const component = parseComponent(file)

  if (component.template) {
    if (component.template.lang === 'pug') {
      try {
        const pug = require('pug')
        component.template.content = pug.render(component.template.content, { filename: resourcePath })
      } catch (err) { /* Ignore compilation errors, they'll be picked up by other loaders */ }
    }
    compile(component.template.content, {
      modules: [{
        postTransformNode: (el) => {
          tags.add(el.tag)
        }
      } as CompilerModuleOptions]
    })
  }

  return [...tags]
}
