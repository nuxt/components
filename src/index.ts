import fs from 'fs'
import path from 'upath'
import chokidar from 'chokidar'
import type { Configuration as WebpackConfig, Entry as WebpackEntry } from 'webpack'
import type { Module } from '@nuxt/types'

import { requireNuxtVersion } from './compatibility'
import { scanComponents } from './scan'
import type { Options, ComponentsDir } from './types'

const isPureObjectOrString = (val: any) => (!Array.isArray(val) && typeof val === 'object') || typeof val === 'string'
const getDir = (p: string) => fs.statSync(p).isDirectory() ? p : path.dirname(p)

const componentsModule: Module<Options> = function () {
  const { nuxt } = this
  const { components } = nuxt.options

  /* istanbul ignore if */
  if (!components) {
    return
  }

  requireNuxtVersion(nuxt?.constructor?.version, '2.10')

  const options: Options = {
    dirs: ['~/components'],
    ...Array.isArray(components) ? { dirs: components } : components
  }

  nuxt.hook('build:before', async (builder: any) => {
    const nuxtIgnorePatterns: string[] = builder.ignore.ignore ? builder.ignore.ignore._rules.map((rule: any) => rule.pattern) : /* istanbul ignore next */ []

    await nuxt.callHook('components:dirs', options.dirs)

    const resolvePath = (dir: any) => nuxt.resolver.resolvePath(dir)

    // Add components/global/ directory
    try {
      const globalDir = getDir(resolvePath('~/components/global'))
      if (!options.dirs.find(dir => resolvePath(dir) === globalDir)) {
        options.dirs.push({
          path: globalDir,
          global: true
        })
      }
    } catch (err) {
      /* istanbul ignore next */
      nuxt.options.watch.push(path.resolve(nuxt.options.srcDir, 'components', 'global'))
    }

    const componentDirs = options.dirs.filter(isPureObjectOrString).map((dir) => {
      const dirOptions: ComponentsDir = typeof dir === 'object' ? dir : { path: dir }

      let dirPath = dirOptions.path
      try { dirPath = getDir(nuxt.resolver.resolvePath(dirOptions.path)) } catch (err) {}

      const transpile = typeof dirOptions.transpile === 'boolean' ? dirOptions.transpile : 'auto'

      // Normalize global option
      if (dirOptions.global === 'dev') {
        dirOptions.global = nuxt.options.dev
      }

      // Normalize level
      dirOptions.level = Number(dirOptions.level || 0)

      const enabled = fs.existsSync(dirPath)
      if (!enabled && dirOptions.path !== '~/components') {
        // eslint-disable-next-line no-console
        console.warn('Components directory not found: `' + dirPath + '`')
      }

      const extensions = dirOptions.extensions || builder.supportedExtensions

      return {
        ...dirOptions,
        enabled,
        path: dirPath,
        extensions,
        pattern: dirOptions.pattern || `**/*.{${extensions.join(',')},}`,
        ignore: [
          '**/*.stories.{js,ts,jsx,tsx}', // ignore storybook files
          ...nuxtIgnorePatterns,
          ...(dirOptions.ignore || [])
        ],
        transpile: (transpile === 'auto' ? dirPath.includes('node_modules') : transpile)
      }
    }).filter(d => d.enabled)

    nuxt.options.build!.transpile!.push(...componentDirs.filter(dir => dir.transpile).map(dir => dir.path))

    let components = await scanComponents(componentDirs, nuxt.options.srcDir!)
    await nuxt.callHook('components:extend', components)

    // Add loader for tree shaking
    if (componentDirs.some(dir => !dir.global)) {
      this.extendBuild((config) => {
        const vueRule = config.module?.rules.find(rule => rule.test?.toString().includes('.vue'))
        if (!vueRule) {
          throw new Error('Cannot find vue loader')
        }
        if (!vueRule.use) {
          vueRule.use = [{
            loader: vueRule.loader!.toString(),
            options: vueRule.options
          }]
          delete vueRule.loader
          delete vueRule.options
        }
        if (!Array.isArray(vueRule!.use)) {
          // @ts-ignore
          vueRule.use = [vueRule.use]
        }

        // @ts-ignore
        vueRule!.use!.unshift({
          loader: require.resolve('./loader'),
          options: {
            dependencies: nuxt.options.dev ? componentDirs.filter(dir => !dir.global).map(dir => dir.path) : /* istanbul ignore next */[],
            getComponents: () => components
          }
        })
      })

      // Add Webpack entry for runtime installComponents function
      nuxt.hook('webpack:config', (configs: WebpackConfig[]) => {
        for (const config of configs.filter(c => ['client', 'modern', 'server'].includes(c.name!))) {
          ((config.entry as WebpackEntry).app as string[]).unshift(path.resolve(__dirname, '../lib/installComponents.js'))
        }
      })
    }

    // Watch
    // istanbul ignore else
    if (nuxt.options.dev && componentDirs.some(dir => dir.watch !== false)) {
      const watcher = chokidar.watch(componentDirs.filter(dir => dir.watch !== false).map(dir => dir.path), nuxt.options.watchers!.chokidar)
      watcher.on('all', async (eventName) => {
        if (!['add', 'unlink'].includes(eventName)) {
          return
        }

        components = await scanComponents(componentDirs, nuxt.options.srcDir!)
        await nuxt.callHook('components:extend', components)

        await builder.generateRoutesAndFiles()
      })

      // Close watcher on nuxt close
      nuxt.hook('close', () => {
        watcher.close()
      })
    }

    // Global components

    // Add templates
    const getComponents = () => components
    const templates = [
      'components/index.js',
      'components/plugin.js',
      'vetur/tags.json'
    ]
    for (const t of templates) {
      this[t.includes('plugin') ? 'addPlugin' : 'addTemplate']({
        src: path.resolve(__dirname, '../templates', t),
        fileName: t,
        options: { getComponents }
      })
    }
  })
}

// @ts-ignore
componentsModule.meta = { name: '@nuxt/components' }

export default componentsModule
