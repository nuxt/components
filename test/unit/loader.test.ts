import { promises as fs } from 'fs'
import { resolve } from 'path'
import { compileToFunctions } from 'vue-template-compiler'
import { loader, DISABLE_COMMENT } from '../../src/loader'
import { scanFixtureComponents } from './utils'

test('loader', async () => {
  const components = await scanFixtureComponents()

  const transform = (code:string) => loader.raw({
    findComponent (name) {
      return components.find(i => i.pascalName === name || i.kebabName === name)
    }
  }, {} as any).transform.call(null, code, '')

  expect(await transform(DISABLE_COMMENT)).toBeFalsy()

  const compiledTemplate = compileToFunctions(`
  <div>
    <Header />
    <Foo />
    <LazyBar />
    <BaseButton />
    <IconHome />
    <MAwesome />
    <Functional />
    <NComponent />
  </div>  
  `).render.toString()
  expect((await transform(compiledTemplate)).code).toMatchSnapshot()
})
