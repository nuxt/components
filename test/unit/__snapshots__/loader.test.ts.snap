// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader 1`] = `
"/* nuxt-components disabled */import __nuxt_components_0 from \\"/Users/antfu/i/nuxt-components/test/fixture/components/Header.vue\\";import __nuxt_components_1 from \\"/Users/antfu/i/nuxt-components/test/fixture/components/Foo.vue\\";import __nuxt_components_2 from \\"/Users/antfu/i/nuxt-components/test/fixture/components/0-base/1.Button.vue\\";import __nuxt_components_3 from \\"/Users/antfu/i/nuxt-components/test/fixture/components/icons/Home.vue\\";import __nuxt_components_4 from \\"/Users/antfu/i/nuxt-components/test/fixture/components/functional/Functional.vue\\";import __nuxt_components_5 from \\"/Users/antfu/i/nuxt-components/test/fixture/components/NComponent.vue\\";function anonymous(
) {
with(this){return _c('div',[_c(__nuxt_components_0),_v(\\" \\"),_c(__nuxt_components_1),_v(\\" \\"),_c('LazyBar'),_v(\\" \\"),_c(__nuxt_components_2),_v(\\" \\"),_c(__nuxt_components_3),_v(\\" \\"),_c('MAwesome'),_v(\\" \\"),_c(__nuxt_components_4),_v(\\" \\"),_c(__nuxt_components_5)],1)}
}"
`;
